### ISSUE TITLE
Improve support for Chrome Android browser

### ISSUE THREAD

**Issue by fajarmf** [collab-project/videojs-record#71]
Currently on Android browser the recording doesn't work. I tried the demo for video and the recorder doesn't start after clicking it.

Any plan on supporting them?

Thank you

**Comment by thijstriemstra** [comment]
@fajarmf what (version) of the android browser? Can you add a screenshot?

**Comment by fajarmf** [comment]
My mistake, it was a chrome browser. and the screenshot is attached.
![screenshot_2016-04-27-08-07-07_com android chrome](https://cloud.githubusercontent.com/assets/191735/14838825/7a0286d2-0c4f-11e6-9170-e40307a26c3d.png)

**Comment by thijstriemstra** [comment]
@fajarmf is there any way you can get a console log for this? I'm hoping we can use that to figure out what's going on here.

**Comment by thijstriemstra** [comment]
I've used the DevTools to debug this issue and I managed to record a audio/video clip but the playback of this clip failed with this error:

```
video.min.js:22 VIDEOJS: ERROR: (CODE:4 MEDIA_ERR_SRC_NOT_SUPPORTED) The media could not be loaded, either because the server or network failed or because the format is not supported. 
```

<img width="1175" alt="screen shot 2016-05-11 at 14 28 50" src="https://cloud.githubusercontent.com/assets/305679/15180990/0152eeb6-1785-11e6-9c6a-2c8be5ad447e.png">

**Comment by thijstriemstra** [comment]
From https://developers.google.com/web/updates/2016/01/mediarecorder: Audio recording works in Firefox and in Chrome 49 and above; Chrome 47 and 48 only support video recording.

I tested audio+video in the previous comment so that explains the error. @fajarmf which one did you try?

**Comment by thijstriemstra** [comment]
Here's a bug tracking the MEDIA_ERR_SRC_NOT_SUPPORTED error: https://bugs.chromium.org/p/chromium/issues/detail?id=253465

There seems to be a [workaround](https://bugs.chromium.org/p/chromium/issues/detail?id=253465#c39), can you give that a try @fajarmf?

**Comment by ignl** [comment]
Hmm for me it also doesn't work on mobile chrome - simply nothing happens. https://webrtc.github.io/samples/src/content/getusermedia/record/ asks for permissions but videojs-record does not. Will try to find out what's wrong later.

**Comment by ignl** [comment]
OK I tried to debug but there are no errors and actually I found out that it kinda works, but reaaaallllyyy slow. Like press button then wait 3 min for video to show up (with pressing that button repeatedly) then press record then wait again and with multiple tries it's possible to record something. So it seems that it does work it just so incredibly slow it's unusable with android chrome. Video is also very slow and choppy while recording.  https://webrtc.github.io/samples/src/content/getusermedia/record/  is really fast compared and original RecordRTC is slower but usable. I have used HTC M8 and android chrome 51.

**Comment by thijstriemstra** [comment]
@ignl thanks for the report. I suspect the code that updates the current time for the video.js player is slow as it happens continuously. What if you comment out the statements on https://github.com/collab-project/videojs-record/blob/master/src/js/videojs.record.js#L1231 and https://github.com/collab-project/videojs-record/blob/master/src/js/videojs.record.js#L1257?

**Comment by ignl** [comment]
Ok I have checked on firefox and video performance seems similar. Maybe chrome is more sensitive to this. I will test it tomorrow as it requires https so i will need to deploy my app or work around that security requirement. However there is also something wrong with the click event as it hardly works on first easy touch on android. For the firefox I also had to try several times until it asked me permission for cam and mic same with chrome just takes much longer. Maybe will need to try and play with different events.

**Comment by ignl** [comment]
I have added touchstart event everywhere where there was a click event. Something like this:
```
DeviceButton = videojs.extend(VjsButton,
    {
        /** @constructor */
        constructor: function(player, options)
        {
            VjsButton.call(this, player, options);

            this.on('touchstart', this.onClick);
            this.on('click', this.onClick);
        }
    });
    DeviceButton.prototype.onClick = function(e)
    {
        e.preventDefault();
        // stop this event before it bubbles up
        e.stopImmediatePropagation();

        // open device dialog
        this.player().recorder.getDevice();
    };
```

Not sure if it is a good approach, I didn't find any problems like ghost click event triggering but maybe its only on my device. Maybe need a better solution like the ones discussed [here](http://stackoverflow.com/questions/7018919/how-to-bind-touchstart-and-click-events-but-not-respond-to-both).

Camera view is laggy and choppy on mobile chrome, and it is so even before starting recording so proposed fix didn't help.  I tried on recordrtc example again and this time recording was also really choppy so I think the problem could be with record rtc.

Also tried on htc browser and it didn't work at all, but the problem is probably that it is based on an old version of chrome so no surprises.

**Comment by thijstriemstra** [comment]
Did adding touch events help?

**Comment by ignl** [comment]
Yes mobile browsers now works fine on my htc just maybe need more testing with more devices. But yeah touch events should be added to the project.

**Comment by thijstriemstra** [comment]
@ignl cool, could you add a PR?

**Comment by ignl** [comment]
Ok I will test my application a bit more, and when finish will try to do it

**Comment by thijstriemstra** [comment]
It looks like the only event video.js subscribes to is the `tap` event as seen [here](https://github.com/videojs/video.js/blob/master/src/js/clickable-component.js#L27). This doesn't seem to be a native browser event but a custom video.js event indicating there is support for touch events as seen [here](https://github.com/videojs/video.js/blob/master/src/js/component.js#L1154). So it seems we need to listen for `tap` events as well as the current `click` events.

**Comment by thijstriemstra** [comment]
@ignl if you have time, please test the latest. I'll put out a new release in a few days.

**Comment by ignl** [comment]
I will test "tap" a bit later today.

**Comment by ignl** [comment]
OK "tap" event works for me.

**Comment by thijstriemstra** [comment]
Did I try it myself? Uh, of course?? And if you followed the link in my previous comment explaining `tap`, you'd see it's a custom video.js event that combines touchstart and touchstop events.

Thanks for testing!

**Comment by ignl** [comment]
Yeah saw "tap" vent before when looking for mobile event, but I thought it will require additional dependency so it's my bad research :)

**Comment by thijstriemstra** [comment]
Going to close this, if there are any other mobile issues, open a new ticket.
